metaflow/plugins/aip/aip.py

```diff
@@ -294,6 +294,72 @@ def _create_workflow_yaml(
 
         KubeflowPipelines._add_archive_section_to_cards_artifacts(workflow)
 
+        if "onExit" in workflow["spec"]:
+            # replace entrypoint content with the exit handler handler content
+            """
+            # What it looks like beforehand...
+            entrypoint: helloflow
+            templates:
+            - name: exit-handler-1
+              dag:
+                tasks:
+                - name: end
+                  template: end
+                  dependencies: [start]
+                - {name: start, template: start}
+            - name: helloflow
+              dag:
+                tasks:
+                - {name: exit-handler-1, template: exit-handler-1}
+                - {name: sqs-exit-handler, template: sqs-exit-handler}
+            """
+            # find the exit-handler-1 template
+            exit_handler_template: dict = [
+                template
+                for template in workflow["spec"]["templates"]
+                if template["name"] == "exit-handler-1"
+            ][0]
+
+            # find the entrypoint template
+            entrypoint_template: dict = [
+                template
+                for template in workflow["spec"]["templates"]
+                if template["name"] == workflow["spec"]["entrypoint"]
+            ][0]
+
+            # replace the entrypoint template with the exit handler template
+            entrypoint_template["dag"] = exit_handler_template["dag"]
+
+            # rename exit-handler-1 to exit-handler
+            exit_handler_template["name"] = "exit-handler"
+            workflow["spec"]["onExit"] = "exit-handler"
+
+            if self.sqs_url_on_error:
+                exit_handler_template["dag"] = {
+                    "tasks": [
+                        {
+                            "name": "sqs-exit-handler",
+                            "template": "sqs-exit-handler",
+                            "when": "{{workflow.status}} != 'Succeeded'",
+                        },
+                    ]
+                }
+
+            if self.notify:
+                notify_task = {
+                    "name": "notify-email-exit-handler",
+                    "template": "notify-email-exit-handler",
+                }
+
+                if self.notify_on_success:
+                    # Always run, even on failure because METAFLOW_NOTIFY_ON_ERROR
+                    # can be injected by the AIP webhook.
+                    pass
+                else:
+                    notify_task["when"] = "{{workflow.status}} != 'Succeeded'"
+
+                exit_handler_template["dag"]["tasks"].append(notify_task)
```

Comment:
Thanks for making these steps conditional! 

I assume if `self.sqs_url_on_error` evaluate to `False`, `exit_handler_template["dag"]` still has the same reference as `entrypoint_template["dag"]`, as it has not been reinitialized in `if self.sqs_url_on_error` block. We can be appending to a different list of tasks. 

Maybe we can instead do
``` python
# initialize
exit_handler_template["dag"] = {"tasks": []}

if self.sqs_url_on_error:
  # append

if self.notify:
  # append
```

Reply:
that is cleaner!

Reply:
fixed!